[Driver] -enable-batch-mode implies -continue-building-after-errors

jrose-apple opened the pull request:
The logic in #16485 to avoid putting certain diagnostics into serialized diagnostic files only makes sense if

1. every diagnostic emitted in file A.swift while processing a different file B.swift would be emitted if we processed A.swift on its own

2. we actually do process A.swift on its own in the same build, or have previously done an incremental build and produced the same diagnostic

But the latter isn't actually guaranteed: if one batch job exits with a failure status, the driver will exit as well, assuming there's no point in continuing. Fortunately, we do have a flag that overrides this behavior, `-continue-building-after-errors`.

(As noted in the patch, `-continue-building-after-errors` isn't *exactly* what we want. But it's conservatively correct.)

jrose-apple commented:
@swift-ci Please smoke test

davidungar commented on lib/Driver/Driver.cpp:
```diff
@@ -703,6 +703,17 @@ Driver::buildCompilation(const ToolChain &TC,
   buildOutputInfo(TC, *TranslatedArgList, BatchMode, Inputs, OI);
 
+  // Note: Batch mode handling of serialized diagnostics requires that all
+  // batches get to run. (This behavior could be limited to only when serialized
```
It would be valuable to explain a bit more about *why* all batches get to run, either here, or somewhere else with a pointer to here. It would save someone a lot of head-scratching someday.

davidungar commented on lib/Driver/Driver.cpp:
```diff
@@ -703,6 +703,17 @@ Driver::buildCompilation(const ToolChain &TC,
+  // to go on to compilation of source files, and if compilation of source files
+  // fails, we shouldn't try to link. Instead, we'd want to let all jobs finish
+  // but not schedule any new ones.
+  ContinueBuildingAfterErrors |= BatchMode;
```
`ContinueBuildingAfterErrors` isn't used before this, so how about moving the earlier code that reads the argument and sets it down to here? That way it could be a `const bool` instead of just a `bool`.

davidungar replied:
Also, consider adding an assertion `ContinueBuildingAfterErrors || !BatchMode` around line 531 in Compilation.cpp
where it says ` return Comp.ContinueBuildingAfterErrors ? ... : ...`

davidungar reviewed (changes requested):
Just a few minor suggestions, but they seem worthwhile to me.

jrose-apple committed:
[Driver] -enable-batch-mode implies -continue-building-after-errors

The logic in 46b8ad3 to avoid putting certain diagnostics into
serialized diagnostic files only makes sense if

1. every diagnostic emitted in file A.swift while processing a
   different file B.swift would be emitted if we processed A.swift on
   its own

2. we actually do process A.swift on its own in the same build,
   or have previously done an incremental build and produced the same
   diagnostic

But the latter isn't actually guaranteed: if one batch job exits with
a failure status, the driver will exit as well, assuming there's no
point in continuing. Fortunately, we do have a flag that overrides
this behavior, -continue-building-after-errors.

(As noted in the patch, -continue-building-after-errors isn't *exactly*
what we want. But it's conservatively correct.)

jrose-apple commented:
Updated!

@swift-ci Please smoke test

davidungar reviewed (approved):
LGTM!
